### PR TITLE
Task 1895 assemble attachments

### PR DESCRIPTION
**Pull request by SRAlexander**

### CODE REVIEW

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+    private XPathService xPathService;
+
+    @Mock
+    private Document ebXmlDocument;
+
+    @InjectMocks
+    private COPCMessageHandler copcMessageHandler;
+
+    @Test
+    public void whenHappyPathThrowNoErrors()
```
Comment (ellenwootten) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:60]: I think that most of the tests have been named `When_HappyPath_Expect_NoErrorsThrown`- we should probably be consistent one way or another going forwards?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+        when(patientAttachmentLogService.findAttachmentLog(inboundMessageId, CONVERSATION_ID))
+            .thenReturn(PatientAttachmentLog.builder()
+                .filename("test_frag_1.txt")
+                .mid("2").parentMid("1")
+                .patientMigrationReqId(1)
+                .build());
+
+        when(patientAttachmentLogService.findAttachmentLogs(CONVERSATION_ID))
+            .thenReturn(new ArrayList<>(
+                    Arrays.asList(
```
Comment (ellenwootten) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:76]: I'd pull this out into a class variable and reuse later on in the file to reduce duplication

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+                                    .compressed(false)
+                                    .contentType("text/plain")
+                                    .lengthNum(0)
+                                    .skeleton(false).build()
+                    )));
+
+        when(attachmentHandlerService.buildSingleFileStringFromPatientAttachmentLogs(any()))
+            .thenReturn("test-string");
+        copcMessageHandler.checkAndMergeFileParts(inboundMessage, CONVERSATION_ID);
+    }
```
Comment (ellenwootten) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:117]: i think it is good practice to have an assertion for each unit test, perhaps even verifying that the mock services you've set up have been called would be better?

**Review by ellenwootten — CHANGES_REQUESTED** [review]
Overall, I think the logic is really solid and is readable. 
There are some stylistic/consistency improvements i've requested/asked about - if there's any issues/questions please let me know.
I've requested changes mainly for the unit test comments - i think they should all have a verify/assert in and would benefit from a tidy up (moving stuff into variables we can reuse in the tests etc)

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java (hunk 1 above)
Comment (rio-k9) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:60]: Check style throws an error when using snake case as asked for. Is the error being suppressed for the formatting in EhrExtractMessageHandlerTest to pass ?

File: db-connector/src/main/resources/uk/nhs/adaptors/connector/dao/PatientAttachmentLogDao/insert-dummy-data.sql [hunk 4]
```diff
@@ -0,0 +1,84 @@
+INSERT INTO patient_attachment_log(
```
Comment (SRAlexander) [inline db-connector/src/main/resources/uk/nhs/adaptors/connector/dao/PatientAttachmentLogDao/insert-dummy-data.sql:1]: Can re rename this file to. attachement-logs-test-data? It's good to have but it's not code critical. Maybe place it in a local folder called local-test-data?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandler.java [hunk 5]
```diff
@@ -21,13 +36,99 @@ public class COPCMessageHandler {
...
+            throw new AttachmentLogException("Given COPC message is missing an attachment log");
+        }
+
+        var conversationAttachmentLogs = patientAttachmentLogService.findAttachmentLogs(conversationId);
+        var attachmentLogFragments = conversationAttachmentLogs.stream()
+            .sorted(Comparator.comparingInt(PatientAttachmentLog::getOrderNum))
+            .filter(log -> log.getParentMid().equals(currentAttachmentLog.getParentMid()))
+            .toList();
+
+        var parentLogMessageId = (attachmentLogFragments.size() == 1)
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandler.java:73]: I think parentheses aren't required here

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java (hunk 3 above)
Comment (mrogowska) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:117]: I think Mockito fails tests which have any unused `when()` statements, so maybe there's no need to verify if a mock service has been called.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/AttachmentHandlerService.java [hunk 6]
```diff
@@ -74,4 +77,29 @@ public void storeAttachments(List<InboundMessage.Attachment> attachments, String
             }
         }
     }
+
+    public byte[] getAttachment(String filename) {
+        if (filename == null || filename.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (rio-k9) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/AttachmentHandlerService.java:82]: replaced with suggested, intellij flagged `StringUtils.isEmpty` deprecated for `StringUtils.hasText`, so used that instead

**Review by ellenwootten — DISMISSED** [review]
Looks good. Good job on the changes. Just a couple of format issues raised by gradle to sort :)

**Review by ellenwootten — APPROVED** [review]

**Pull request merged**